BKMoneyKit/BKBirthdayField.m

```diff
@@ -0,0 +1,242 @@
+                                                                      range:NSMakeRange(0, range.location)];
+        // adjust replace range
+        if (numberCharacterRange.location != NSNotFound) {
+            range = NSUnionRange(range, numberCharacterRange);
+        }
+    }
+    
+    NSString *replacedString = [currentText stringByReplacingCharactersInRange:range withString:string];
+    NSString *numberOnlyString = [self numberOnlyStringWithString:replacedString];
+    
+    if (numberOnlyString.length > 8) {
+        return NO;
+    }
+    
+    if (numberOnlyString.length == 1 && [numberOnlyString substringToIndex:1].integerValue > 1) {
+        numberOnlyString = [@"0" stringByAppendingString:numberOnlyString];
+    }
+    
+    NSInteger monthInteger = 0;
+    NSMutableString *formattedString = [NSMutableString string];
+    
+    if (numberOnlyString.length > 0) {
+        NSString *monthString = [numberOnlyString substringToIndex:MIN(2, numberOnlyString.length)];
+        
+        if (monthString.length == 2) {
+            monthInteger = monthString.integerValue;
+            if (monthInteger < 1 || monthInteger > 12) {
+                return NO;
+            }
+        }
+        [formattedString appendString:monthString];
+    }
+    
+    if (numberOnlyString.length > 1) {
+        [formattedString appendString:@" / "];
+    }
+    
+    if ([numberOnlyString length] > 2) {
+        NSInteger numOfDayDigits = numberOnlyString.length > 3 ? 2 : 1;
+        NSString *dayString = [numberOnlyString substringWithRange:NSMakeRange(2, numOfDayDigits)];
+        
+        if (dayString.length == 1) {
+            NSInteger dayInteger = dayString.integerValue;
+            
+            if (dayInteger < 0 || dayInteger > 3) {
+                return NO;
+            }
+        } else if (dayString.length == 2) {
+            NSInteger dayInteger = dayString.integerValue;
+            
+            switch (monthInteger) {
+                case 1:
+                case 3:
+                case 5:
+                case 7:
+                case 8:
+                case 10:
+                case 12:
+                    if (dayInteger < 1 || dayInteger > 31) {
+                        return NO;
+                    }
+                    break;
+                case 4:
+                case 6:
+                case 9:
+                case 11:
+                    if (dayInteger < 1 || dayInteger > 30) {
+                        return NO;
+                    }
+                    break;
+                case 2:
+                    // Month is typed before year, won't know if it's a leap year, bound check against upper limit
+                    if (dayInteger < 1 || dayInteger > 29) {
+                        return NO;
+                    }
+                    break;
+                default:
+                    return NO;
+                    break;
+            }
```

Comment:
Hello @bkook:

Problem with trying to use NSDateComponents is that with the MM / DD / YYYY format, the year hasn't been typed in yet so I can't use NSDateComponents to do the actual calculations. And generally it would work, except for leap months.

I also feel weird instantiating NSDate objects and using NSCalendar to do pretty heavy math when the current solution gets the correct date in just 3 operations (1 switch and 2 integer comparisons). I'll move that code to a separate method so it reads cleaner.